File: dart-sdks/test/sui/multisig_test.dart
import 'dart:convert';

import 'package:dart_sdks/dart_sdks.dart';
import 'package:test/test.dart';

void main() {
  test('threshold verification', () async {
    final k1 = await Ed25519Keypair.generate();
    final k2 = await Ed25519Keypair.generate();
    final k3 = await Ed25519Keypair.generate();

    final message = utf8.encode('m');
    final pub = MultisigPublicKey(
      publicKeys: [await k1.publicKeyBytes(), await k2.publicKeyBytes(), await k3.publicKeyBytes()],
      weights: [1, 1, 1],
      threshold: 2,
    );

    final signer = MultisigSigner(pub, scheme: SignatureScheme.ed25519);

    final ms = MultisigSignature(
      signatures: [await k1.sign(message), await k2.sign(message)],
      bitmap: [0, 1],
    );
    expect(await signer.verify(message, ms), isTrue);

    final low = MultisigSignature(signatures: [await k1.sign(message)], bitmap: [0]);
    expect(await signer.verify(message, low), isFalse);
  });

  test('duplicate index rejected', () async {
    final k1 = await Ed25519Keypair.generate();
    final message = utf8.encode('m');
    final pub = MultisigPublicKey(
      publicKeys: [await k1.publicKeyBytes()],
      weights: [2],
      threshold: 2,
    );
    final signer = MultisigSigner(pub, scheme: SignatureScheme.ed25519);
    final ms = MultisigSignature(
      signatures: [await k1.sign(message), await k1.sign(message)],
      bitmap: [0, 0],
    );
    expect(await signer.verify(message, ms), isFalse);
  });

  test('signature roundtrip serialization', () {
    final ms = MultisigSignature(signatures: [utf8.encode('sig1'), utf8.encode('sig2')], bitmap: [0, 2]);
    final encoded = ms.toBase64();
    final decoded = MultisigSignature.fromBase64(encoded);
    expect(decoded.bitmap, [0, 2]);
    expect(decoded.signatures.length, 2);
  });

  test('build and sign helpers', () async {
    final k1 = await Ed25519Keypair.generate();
    final k2 = await Ed25519Keypair.generate();
    final k3 = await Ed25519Keypair.generate();
    final message = utf8.encode('m');

    final pub = MultisigPublicKey(
      publicKeys: [await k1.publicKeyBytes(), await k2.publicKeyBytes(), await k3.publicKeyBytes()],
      weights: [1, 1, 1],
      threshold: 2,
    );
    final signer = MultisigSigner(pub, scheme: SignatureScheme.ed25519);

    final built = await signer.build(
      message,
      [
        (0, await k1.sign(message)),
        (1, await k2.sign(message)),
      ],
    );
    expect(built.bitmap, [0, 1]);
    expect(signer.isThresholdMet(built), isTrue);

    final signed = await signer.sign(
      message,
      [
        (1, (msg) => k2.sign(msg)),
        (2, (msg) => k3.sign(msg)),
      ],
    );
    expect(signed.bitmap, [1, 2]);
  });

  test('build validation errors', () async {
    final k1 = await Ed25519Keypair.generate();
    final k2 = await Ed25519Keypair.generate();
    final message = utf8.encode('m');

    final pub = MultisigPublicKey(
      publicKeys: [await k1.publicKeyBytes(), await k2.publicKeyBytes()],
      weights: [1, 1],
      threshold: 2,
    );
    final signer = MultisigSigner(pub, scheme: SignatureScheme.ed25519);

    await expectLater(signer.build(message, []), throwsArgumentError);

    await expectLater(
      signer.build(
        message,
        [
          (0, await k1.sign(message)),
          (0, await k2.sign(message)),
        ],
      ),
      throwsArgumentError,
    );
  });
}

File: rust-sdks/crates/sui/tests/multisig_test.rs
use base64::Engine as _;
use sui::crypto::SignatureScheme;
use sui::keypairs::ed25519;
use sui::multisig::{
    parse_multisig, serialize_multisig, MultiSigEntry, MultiSigPublicKey, MultiSigSerialized,
    WeightedPublicKey,
};

#[test]
fn multisig_serialize_parse_and_verify() {
    let k1 = ed25519::Keypair::generate();
    let k2 = ed25519::Keypair::generate();

    let pk1 = base64::engine::general_purpose::STANDARD.encode(k1.public_key_bytes());
    let pk2 = base64::engine::general_purpose::STANDARD.encode(k2.public_key_bytes());

    let ms = MultiSigPublicKey {
        public_keys: vec![
            WeightedPublicKey {
                scheme: SignatureScheme::Ed25519,
                public_key: pk1.clone(),
                weight: 1,
            },
            WeightedPublicKey {
                scheme: SignatureScheme::Ed25519,
                public_key: pk2.clone(),
                weight: 1,
            },
        ],
        threshold: 2,
    };

    let msg = b"hello-multisig";
    let sig1 = base64::engine::general_purpose::STANDARD.encode(k1.sign(msg));
    let sig2 = base64::engine::general_purpose::STANDARD.encode(k2.sign(msg));

    let serialized = serialize_multisig(&MultiSigSerialized {
        signatures: vec![
            MultiSigEntry {
                public_key: pk1,
                signature: sig1,
            },
            MultiSigEntry {
                public_key: pk2,
                signature: sig2,
            },
        ],
        bitmap: vec![0, 1],
        threshold: 2,
    })
    .expect("serialize");

    let parsed = parse_multisig(&serialized).expect("parse");
    assert_eq!(parsed.signatures.len(), 2);

    let ok = ms.verify(msg, &serialized).expect("verify multisig");
    assert!(ok);
}
